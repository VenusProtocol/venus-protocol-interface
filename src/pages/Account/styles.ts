import { css } from '@emotion/react';
import { useTheme } from '@mui/material';

export const useStyles = () => {
  const theme = useTheme();

  return {
    section: css`
      :not(:last-child) {
        margin-bottom: ${theme.spacing(14)};
      }

      ${theme.breakpoints.down('lg')} {
        :not(:last-child) {
          margin-bottom: ${theme.spacing(10)};
        }
      }
    `,
    sectionTitle: css`
      display: flex;
      align-items: center;
      justify-content: space-between;
      margin-bottom: ${theme.spacing(6)};

      ${theme.breakpoints.down('lg')} {
        margin-bottom: ${theme.spacing(4)};
      }

      ${theme.breakpoints.down('md')} {
        flex-direction: column-reverse;
        align-items: flex-start;
        margin-bottom: ${theme.spacing(4)};
      }
    `,
    sectionTitleText: css`
      ${theme.breakpoints.down('md')} {
        font-size: ${theme.typography.h4.fontSize};
        font-weight: ${theme.typography.h4.fontWeight};
      }
    `,
    sectionTitleToggle: css`
      ${theme.breakpoints.down('md')} {
        margin-bottom: ${theme.spacing(6)};
      }
    `,
    getNetApyColor: ({ netApyPercentage }: { netApyPercentage: number }) =>
      netApyPercentage >= 0 ? theme.palette.interactive.success : theme.palette.interactive.error,
  };
};
