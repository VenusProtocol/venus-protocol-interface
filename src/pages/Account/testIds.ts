const TEST_IDS = {
  stats: 'account-stats',
};

export default TEST_IDS;
