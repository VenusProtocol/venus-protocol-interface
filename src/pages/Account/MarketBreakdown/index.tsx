/** @jsxImportSource @emotion/react */
import { Paper, Typography } from '@mui/material';
import BigNumber from 'bignumber.js';
import {
  BorrowLimitUsedAccountHealth,
  Cell,
  CellGroup,
  Icon,
  RiskLevel,
  Tooltip,
} from 'components';
import React from 'react';
import { useTranslation } from 'translation';
import { Asset, MarketRiskLevel } from 'types';
import { formatCentsToReadableValue, formatToReadablePercentage } from 'utilities';

import { SAFE_BORROW_LIMIT_PERCENTAGE } from 'constants/safeBorrowLimitPercentage';

import Tables from './Tables';
import { useStyles } from './styles';
import TEST_IDS from './testIds';
import useExtractData from './useExtractData';

export interface MarketBreakdownProps {
  marketName: string;
  riskLevel: MarketRiskLevel;
  isXvsEnabled: boolean;
  assets: Asset[];
  dailyXvsDistributionInterestsCents?: BigNumber;
  className?: string;
}

// TODO: add tests

export const MarketBreakdown: React.FC<MarketBreakdownProps> = ({
  marketName,
  assets,
  isXvsEnabled,
  riskLevel,
  dailyXvsDistributionInterestsCents,
  className,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const {
    totalSupplyCents,
    totalBorrowCents,
    borrowLimitCents,
    readableSafeBorrowLimit,
    safeBorrowLimitPercentage,
    dailyEarningsCents,
    netApyPercentage,
  } = useExtractData({
    assets,
    isXvsEnabled,
    dailyXvsDistributionInterestsCents,
  });

  const cells: Cell[] = [
    {
      label: t('account.marketBreakdown.cellGroup.netApy'),
      value: formatToReadablePercentage(netApyPercentage),
      tooltip: t('account.marketBreakdown.cellGroup.netApyTooltip'),
      color: styles.getNetApyColor({ netApyPercentage: netApyPercentage || 0 }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.dailyEarnings'),
      value: formatCentsToReadableValue({ value: dailyEarningsCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalSupply'),
      value: formatCentsToReadableValue({ value: totalSupplyCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalBorrow'),
      value: formatCentsToReadableValue({ value: totalBorrowCents }),
    },
  ];

  return (
    <div className={className}>
      <div css={styles.title}>
        <Typography css={styles.marketName} variant="h4">
          {marketName}
        </Typography>

        <RiskLevel variant={riskLevel} />
      </div>

      <Paper css={styles.statsContainer} data-testid={TEST_IDS.stats}>
        <CellGroup smallValues cells={cells} css={styles.cellGroup} />

        <div css={styles.accountHealth}>
          <BorrowLimitUsedAccountHealth
            variant="borrowLimitUsed"
            borrowBalanceCents={totalBorrowCents.toNumber()}
            borrowLimitCents={borrowLimitCents.toNumber()}
            safeBorrowLimitPercentage={SAFE_BORROW_LIMIT_PERCENTAGE}
            css={styles.accountHealthProgressBar}
          />

          <div css={styles.accountHealthFooter}>
            <Icon name="shield" css={styles.shieldIcon} />

            <Typography component="span" variant="small2" css={styles.inlineLabel}>
              {t('myAccount.safeLimit')}
            </Typography>

            <Typography
              component="span"
              variant="small1"
              color="text.primary"
              css={styles.safeLimit}
            >
              {readableSafeBorrowLimit}
            </Typography>

            <Tooltip
              css={styles.tooltip}
              title={t('myAccount.safeLimitTooltip', { safeBorrowLimitPercentage })}
            >
              <Icon css={styles.infoIcon} name="info" />
            </Tooltip>
          </div>
        </div>
      </Paper>

      <Tables assets={assets} isXvsEnabled={isXvsEnabled} />
    </div>
  );
};

export default MarketBreakdown;
