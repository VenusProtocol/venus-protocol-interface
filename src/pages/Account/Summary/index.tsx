/** @jsxImportSource @emotion/react */
import { Paper, Typography } from '@mui/material';
import { BorrowLimitUsedAccountHealth, Cell, CellGroup, Icon, Tooltip } from 'components';
import React from 'react';
import { useTranslation } from 'translation';
import { Asset } from 'types';
import { formatCentsToReadableValue, formatToReadablePercentage } from 'utilities';

import { useStyles } from './styles';
import TEST_IDS from './testIds';
import useExtractData from './useExtractData';

export interface SummaryProps {
  assets: Asset[];
  safeBorrowLimitPercentage?: number;
  displayAccountHealth?: boolean;
  className?: string;
}

export const Summary: React.FC<SummaryProps> = ({
  assets,
  displayAccountHealth,
  safeBorrowLimitPercentage,
  className,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const {
    totalSupplyCents,
    totalBorrowCents,
    borrowLimitCents,
    readableSafeBorrowLimit,
    dailyEarningsCents,
    netApyPercentage,
  } = useExtractData({
    assets,
    safeBorrowLimitPercentage,
  });

  const cells: Cell[] = [
    {
      label: t('account.marketBreakdown.cellGroup.netApy'),
      value: formatToReadablePercentage(netApyPercentage),
      tooltip: t('account.marketBreakdown.cellGroup.netApyTooltip'),
      color: styles.getNetApyColor({ netApyPercentage: netApyPercentage || 0 }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.dailyEarnings'),
      value: formatCentsToReadableValue({ value: dailyEarningsCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalSupply'),
      value: formatCentsToReadableValue({ value: totalSupplyCents }),
    },
    {
      label: t('account.marketBreakdown.cellGroup.totalBorrow'),
      value: formatCentsToReadableValue({ value: totalBorrowCents }),
    },
  ];

  return (
    <Paper css={styles.container} className={className} data-testid={TEST_IDS.container}>
      <CellGroup
        smallValues={displayAccountHealth}
        cells={cells}
        css={styles.cellGroup}
        data-testid={TEST_IDS.stats}
      />

      {displayAccountHealth && safeBorrowLimitPercentage && (
        <div css={styles.accountHealth} data-testid={TEST_IDS.accountHealth}>
          <BorrowLimitUsedAccountHealth
            variant="borrowLimitUsed"
            borrowBalanceCents={totalBorrowCents.toNumber()}
            borrowLimitCents={borrowLimitCents.toNumber()}
            safeBorrowLimitPercentage={safeBorrowLimitPercentage}
            css={styles.accountHealthProgressBar}
          />

          <div css={styles.accountHealthFooter}>
            <Icon name="shield" css={styles.shieldIcon} />

            <Typography component="span" variant="small2" css={styles.inlineLabel}>
              {t('myAccount.safeLimit')}
            </Typography>

            <Typography
              component="span"
              variant="small1"
              color="text.primary"
              css={styles.safeLimit}
            >
              {readableSafeBorrowLimit}
            </Typography>

            <Tooltip
              css={styles.tooltip}
              title={t('myAccount.safeLimitTooltip', { safeBorrowLimitPercentage })}
            >
              <Icon css={styles.infoIcon} name="info" />
            </Tooltip>
          </div>
        </div>
      )}
    </Paper>
  );
};

export default Summary;
