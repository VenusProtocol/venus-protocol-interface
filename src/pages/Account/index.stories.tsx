import { ComponentMeta } from '@storybook/react';

import { withRouter } from 'stories/decorators';

import Account from '.';

export default {
  title: 'Pages/Account',
  component: Account,
  decorators: [withRouter],
} as ComponentMeta<typeof Account>;

export { Account };
