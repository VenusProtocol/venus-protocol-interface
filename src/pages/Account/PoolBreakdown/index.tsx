/** @jsxImportSource @emotion/react */
import { Typography } from '@mui/material';
import React from 'react';
import { Pool } from 'types';
import { isFeatureEnabled } from 'utilities';

import Summary from '../Summary';
import Tables from './Tables';
import { useStyles } from './styles';
import TEST_IDS from './testIds';

export interface PoolBreakdownProps {
  pool: Pool;
  className?: string;
}

export const PoolBreakdown: React.FC<PoolBreakdownProps> = ({ pool, className }) => {
  const styles = useStyles();

  return (
    <div className={className}>
      {isFeatureEnabled('isolatedPools') && (
        <div css={styles.title} data-testid={TEST_IDS.title}>
          <Typography css={styles.marketName} variant="h3">
            {pool.name}
          </Typography>
        </div>
      )}

      <Summary pools={[pool]} displayAccountHealth css={styles.summary} />

      <Tables pool={pool} />
    </div>
  );
};

export default PoolBreakdown;
