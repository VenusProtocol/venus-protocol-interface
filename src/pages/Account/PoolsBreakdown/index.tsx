/** @jsxImportSource @emotion/react */
import { ProgressCircle, Tag, TagGroup, Tooltip } from 'components';
import React, { useMemo, useState } from 'react';
import { useTranslation } from 'translation';
import { Pool } from 'types';
import { calculatePercentage, formatPercentageToReadableValue, isFeatureEnabled } from 'utilities';

import Section from '../Section';
import Summary from '../Summary';
import Tables from './Tables';
import { useStyles } from './styles';

export interface PoolsBreakdownProps {
  pools: Pool[];
  className?: string;
}

export const PoolsBreakdown: React.FC<PoolsBreakdownProps> = ({ pools, className }) => {
  const styles = useStyles();
  const { t } = useTranslation();
  const [selectedPoolIndex, setSelectedPoolIndex] = useState<number>(0);
  const selectedPool = pools[selectedPoolIndex];

  const tags: Tag[] = useMemo(
    () =>
      pools.map(pool => {
        const borrowLimitUsedPercentage =
          pool.userBorrowBalanceCents &&
          pool.userBorrowLimitCents &&
          calculatePercentage({
            numerator: pool.userBorrowBalanceCents.toNumber(),
            denominator: pool.userBorrowLimitCents.toNumber(),
          });

        const readableBorrowLimitUsedPercentage =
          formatPercentageToReadableValue(borrowLimitUsedPercentage);

        return {
          id: pool.comptrollerAddress,
          content: (
            <>
              <span>{pool.name}</span>

              {borrowLimitUsedPercentage !== undefined && (
                <Tooltip
                  title={t('account.poolsBreakdown.poolTagTooltip', {
                    borrowLimitUsedPercentage: readableBorrowLimitUsedPercentage,
                  })}
                  css={styles.tagTooltip}
                >
                  <ProgressCircle value={borrowLimitUsedPercentage} />
                </Tooltip>
              )}
            </>
          ),
        };
      }),
    [pools],
  );

  return (
    <Section
      className={className}
      title={
        isFeatureEnabled('isolatedPools') ? t('account.poolsBreakdown.title') : selectedPool.name
      }
    >
      {isFeatureEnabled('isolatedPools') && pools.length > 0 && (
        <TagGroup
          css={styles.tags}
          tags={tags}
          activeTagIndex={selectedPoolIndex}
          onTagClick={setSelectedPoolIndex}
        />
      )}

      <Summary pools={[selectedPool]} displayAccountHealth css={styles.summary} />

      <Tables pool={selectedPool} />
    </Section>
  );
};

export default PoolsBreakdown;
