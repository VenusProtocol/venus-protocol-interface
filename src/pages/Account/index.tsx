/** @jsxImportSource @emotion/react */
import { Typography } from '@mui/material';
import { Cell, CellGroup, Toggle } from 'components';
import React, { useContext } from 'react';
import { useTranslation } from 'translation';
import { Pool } from 'types';
import { formatCentsToReadableValue, formatToReadablePercentage } from 'utilities';

import { poolData } from '__mocks__/models/pools';
import { IncludeXvsContext } from 'context/IncludeXvsContext';

import PoolBreakdown from './PoolBreakdown';
import { useStyles } from './styles';
import TEST_IDS from './testIds';

export interface AccountUiProps {
  pools: Pool[];
  netApyPercentage: number;
  dailyEarningsCents: number;
  totalSupplyCents: number;
  totalBorrowCents: number;
  includeXvs: boolean;
  onIncludeXvsToggleChange: (newValue: boolean) => void;
}

export const AccountUi: React.FC<AccountUiProps> = ({
  netApyPercentage,
  dailyEarningsCents,
  totalSupplyCents,
  totalBorrowCents,
  includeXvs,
  pools,
  onIncludeXvsToggleChange,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const cells: Cell[] = [
    {
      label: t('account.accountSummary.netApy'),
      value: formatToReadablePercentage(netApyPercentage),
      tooltip: t('account.accountSummary.netApyTooltip'),
      color: styles.getNetApyColor({ netApyPercentage }),
    },
    {
      label: t('account.accountSummary.dailyEarnings'),
      value: formatCentsToReadableValue({ value: dailyEarningsCents }),
    },
    {
      label: t('account.accountSummary.totalSupply'),
      value: formatCentsToReadableValue({ value: totalSupplyCents }),
    },
    {
      label: t('account.accountSummary.totalBorrow'),
      value: formatCentsToReadableValue({ value: totalBorrowCents }),
    },
  ];

  return (
    <>
      <div css={styles.section}>
        <div css={styles.sectionTitle}>
          <Typography variant="h3" css={styles.sectionTitleText}>
            {t('account.accountSummary.title')}
          </Typography>

          <Toggle
            css={styles.sectionTitleToggle}
            tooltip={t('account.accountSummary.includeXvsToggleTooltip')}
            label={t('account.accountSummary.includeXvsToggleLabel')}
            isLight
            value={includeXvs}
            onChange={event => onIncludeXvsToggleChange(event.currentTarget.checked)}
          />
        </div>

        <CellGroup cells={cells} data-testid={TEST_IDS.stats} />
      </div>

      {pools.map(pool => (
        <PoolBreakdown
          key={`pool-breakdown-${pool.name}`}
          css={styles.section}
          pool={pool}
          includeXvs={includeXvs}
        />
      ))}
    </>
  );
};

const Account: React.FC = () => {
  const { includeXvs, setIncludeXvs } = useContext(IncludeXvsContext);

  // TODO: fetch (see VEN-548)
  const netApyPercentage = 13.4;
  const dailyEarningsCents = 100000;
  const totalSupplyCents = 100000000;
  const totalBorrowCents = 10000000;

  return (
    <AccountUi
      netApyPercentage={netApyPercentage}
      dailyEarningsCents={dailyEarningsCents}
      totalSupplyCents={totalSupplyCents}
      totalBorrowCents={totalBorrowCents}
      includeXvs={includeXvs}
      onIncludeXvsToggleChange={setIncludeXvs}
      pools={poolData}
    />
  );
};

export default Account;
