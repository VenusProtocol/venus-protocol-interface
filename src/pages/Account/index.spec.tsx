import React from 'react';

import renderComponent from 'testUtils/renderComponent';

import Account from '.';

jest.mock('clients/api');

describe('pages/Account', () => {
  it('renders without crashing', () => {
    renderComponent(<Account />);
  });

  // TODO: add tests
});
