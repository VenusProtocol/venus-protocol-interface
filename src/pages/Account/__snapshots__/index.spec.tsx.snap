// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Account displays stats correctly 1`] = `"Net APY13.4%Daily Earnings$1,000.00Total Supply$1,000,000.00Total Borrow$100,000.00"`;
